packages/@aws-cdk/aws-sagemaker/lib/model.ts

```diff
@@ -0,0 +1,405 @@
+
+/**
+ * Name tag constant
+ */
+const NAME_TAG: string = 'Name';
+
+/**
+ * Interface that defines a Model resource.
+ */
+export interface IModel extends cdk.IResource, iam.IGrantable, ec2.IConnectable {
+  /**
+   * Returns the ARN of this model.
+   *
+   * @attribute
+   */
+  readonly modelArn: string;
+
+  /**
+   * Returns the name of this model.
+   *
+   * @attribute
+   */
+  readonly modelName: string;
+
+  /**
+   * The IAM role associated with this Model.
+   */
+  readonly role?: iam.IRole;
+
+  /**
+   * Adds a statement to the IAM role assumed by the instance.
+   */
+  addToRolePolicy(statement: iam.PolicyStatement): void;
+}
+
+/**
+ * Represents a Model resource defined outside this stack.
+ */
+export interface ModelAttributes {
+  /**
+   * The name of this model.
+   */
+  readonly modelName: string;
+
+  /**
+   * The IAM execution role associated with this model.
+   *
+   * @default - When not provided, any role-related operations will no-op.
+   */
+  readonly role?: iam.IRole;
+
+  /**
+   * The security groups for this model, if in a VPC.
+   *
+   * @default - When not provided, the connections to/from this model cannot be managed.
+   */
+  readonly securityGroups?: ec2.ISecurityGroup[];
+}
+
+abstract class ModelBase extends cdk.Resource implements IModel {
+  /**
+   * Returns the ARN of this model.
+   * @attribute
+   */
+  public abstract readonly modelArn: string;
+  /**
+   * Returns the name of the model.
+   * @attribute
+   */
+  public abstract readonly modelName: string;
+  /**
+   * Execution role for SageMaker Model
+   */
+  public abstract readonly role?: iam.IRole;
+  /**
+   * The principal this Model is running as
+   */
+  public abstract readonly grantPrincipal: iam.IPrincipal;
+  /**
+   * An accessor for the Connections object that will fail if this Model does not have a VPC
+   * configured.
+   */
+  public get connections(): ec2.Connections {
+    if (!this._connections) {
+      throw new Error('Cannot manage network access without configuring a VPC');
+    }
+    return this._connections;
+  }
+  /**
+   * The actual Connections object for this Model. This may be unset in the event that a VPC has not
+   * been configured.
+   * @internal
+   */
+  protected _connections: ec2.Connections | undefined;
+
+  /**
+   * Adds a statement to the IAM role assumed by the instance.
+   */
+  public addToRolePolicy(statement: iam.PolicyStatement) {
+    if (!this.role) {
+      return;
+    }
+
+    this.role.addToPolicy(statement);
+  }
+}
+
+/**
+ * Describes the container, as part of model definition.
+ */
+export interface ContainerDefinition {
+  /**
+   * The image used to start a container.
+   */
+  readonly image: ContainerImage;
+
+  /**
+   * A map of environment variables to pass into the container.
+   *
+   * @default - none
+   */
+  readonly environment?: {[key: string]: string};
+
+  /**
+   * Hostname of the container.
+   *
+   * @default - none
```

Comment:
the default is no hostname?

Reply:
I think we should do 3) and just document the details and/or link to the doc. That's enough for this initial PR.

Future work could include adding a warning annotation for single containers with hostname specified (a courtesy warning saying that the hostname is ignored). We could design a synth-time exception for partially specific hostnames as well, but that is definitely one of those `p2` nice-to-haves that can be done down the line.